vowpalwabbit/offset_tree.cc

```diff
@@ -0,0 +1,296 @@
+
+  uint32_t min_depth_binary_tree::leaf_node_count() const { return _num_leaf_nodes; }
+
+  void offset_tree::init(uint32_t num_actions) { binary_tree.build_tree(num_actions); }
+
+  int32_t offset_tree::learner_count() const { return binary_tree.internal_node_count(); }
+
+  // Helper to deal with collections that don't start with an index of 0
+  template<typename T>
+  struct offset_helper
+  {
+    // typedef verbose prediction buffer type
+    offset_helper(T& b, uint32_t index_offset)
+    : start_index_offset{index_offset}, collection(b) {}
+
+    // intercept index operator to adjust the offset before
+    // passing to underlying collection
+    typename T::const_reference operator[](size_t idx) const
+    {
+      return collection[idx - start_index_offset];
+    }
+
+    typename T::reference operator[](size_t idx) {
+      return collection[idx - start_index_offset];
+    }
+
+  private:
+    uint32_t start_index_offset = 0;
+    T& collection;
+  };
+
+  const offset_tree::scores_t& offset_tree::predict(LEARNER::single_learner& base, example& ec)
+  {
+    // - pair<float,float> stores the scores for left and right nodes
+    // - prediction_buffer stores predictions for all the nodes in the tree for the duration
+    //   of the predict() call
+    // - static thread_local ensures only one copy per calling thread.  This is to reduce
+    //   memory allocations in steady state.
+    using predict_buffer_t = std::vector<std::pair<float, float>>;
+    static thread_local predict_buffer_t prediction_buffer(learner_count());
+    static thread_local std::vector<float> scores(binary_tree.leaf_node_count());
+
+    auto& t = binary_tree;
+
+    prediction_buffer.clear();
+    scores.resize(t.leaf_node_count());
+
+    // Handle degenerate cases of zero and one node trees
+    if (t.leaf_node_count() == 0)
+      return scores;
+    if (t.leaf_node_count() == 1)
+    {
+      scores[0] = 1.0f;
+      return scores;
+    }
+
+    // Save and clear example label.
+    // (CB algs predict_or_learn checks for valid label during learn.)
+    static thread_local CB::label saved_label;
+    saved_label = ec.l.cb;
+    ec.l.cb.costs.clear();
+
+    // Get predictions for all internal nodes
+    for (uint32_t idx = 0; idx < t.internal_node_count(); ++idx)
+    {
+      base.predict(ec, idx);
+      prediction_buffer.emplace_back(ec.pred.a_s[0].score, ec.pred.a_s[1].score);
+    }
+
+    // Restore example label.
+    ec.l.cb = saved_label;
+
+    // use a offset helper to deal with start index offset
+    offset_helper<predict_buffer_t> buffer_helper(prediction_buffer, t.leaf_node_count());
+
+    // Compute action scores
+    for (auto rit = t.nodes.rbegin(); rit != t.nodes.rend(); ++rit)
+    {
+      // done processing all internal nodes
+      if (rit->is_leaf)
+        break;
+
+      // update probabilities for left node
+      const float left_p = buffer_helper[rit->id].first;
```

Comment:
should the probability updates for left and right node be refactored into a method call to avoid duplication here?

Reply:
They touch left_id or right_id and first or second.  This seems clearer to me.